# Normalization

Import libraries and functions.

In [1]:
import pandas as pd
import numpy as np
import glob
import os
from zipfile import ZipFile
import warnings
warnings.filterwarnings("ignore")
import functools as ft
import ipywidgets as widgets
from ipywidgets import Layout
from ipywidgets import interact, interact_manual
from pandas.api.types import is_numeric_dtype

Taking as reference both works of https://www.pluralsight.com/guides/cleaning-up-data-from-outliers and https://careerfoundry.com/en/blog/data-analytics/how-to-find-outliers/, for normalizing our data we need to start computing the outliers and removing them from our dataframe. As there is not a direct function of pandas that performs this step, it´s been step-by-step code, where we begin with the computation of the quartiles, then the IQR (Inter Quartile Range) and finally the upper and lower limit.

##### IQR explanation

The interquartile range (IQR) measures the spread of the middle half of your data. It is the range for the middle 50% of your sample. Use the IQR to assess the variability where most of your values lie. Larger values indicate that the central portion of your data spread out further. Conversely, smaller values show that the middle values cluster more tightly.

To visualize the interquartile range, imagine dividing your data into quarters. Statisticians refer to these quarters as quartiles and label them from low to high as Q1, Q2, Q3, and Q4. The lowest quartile (Q1) covers the smallest quarter of values in your dataset. The upper quartile (Q4) comprises the highest quarter of values. The interquartile range is the middle half of the data that lies between the upper and lower quartiles. In other words, the interquartile range includes the 50% of data points that are above Q1 and below Q4. The IQR is the red area in the graph below, containing Q2 and Q3 (not labeled).

https://camo.githubusercontent.com/a5f6cf8164048f8c28f9b00b94e1264480c8c3b20a1b3d0bdca47083f3a86a19/68747470733a2f2f69302e77702e636f6d2f7374617469737469637362796a696d2e636f6d2f77702d636f6e74656e742f75706c6f6164732f323031382f30332f696e7465727175617274696c655f72616e67652e706e673f773d3537362673736c3d31

When measuring variability, statisticians prefer using the interquartile range instead of the full data range because extreme values and outliers affect it less. Typically, use the IQR with a measure of central tendency, such as the median, to understand your data’s center and spread. This combination creates a fuller picture of your data’s distribution.

Therefore it is being utilized to get rid of all the outliers that may come from errors when creating the data or from unexpected years.

Firstly, we compute the first quartile (Q1=25%) and the third quartile (Q3=75%). For that, we have grouped the data by country code and indicator name, so we get the Q1 and Q3 values for each indicator in each geographical area. 

In [11]:
grouped=BronzeDataFrame.groupby(['Country Code','Indicator Name'])
grouped

In [12]:
Q1=BronzeDataFrame.groupby(['Country Code','Indicator Name']).quantile(0.25)
Q3=BronzeDataFrame.groupby(['Country Code','Indicator Name']).quantile(0.75)
IQR=Q3-Q1
IQR

Date  \
Country Code Indicator Name                                             
ARE          Access to clean fuels and technologies for cook...  10.0   
             Access to clean fuels and technologies for cook...  10.0   
             Access to clean fuels and technologies for cook...  10.0   
             Access to electricity (% of population)             15.0   
             Access to electricity, rural (% of rural popula...  15.0   
...                                                               ...   
ZAF          Women who believe a husband is justified in bea...   0.0   
             Women who were first married by age 15 (% of wo...   9.0   
             Women who were first married by age 18 (% of wo...   9.0   
             Women's share of population ages 15+ living wit...  15.0   
             Young people (ages 15-24) newly infected with HIV   15.0   

                                                                    Value  
Country Code Indicator Name                                                
ARE          Access to clean fuels and technologies for cook...      0.00  
             Access to clean fuels and technologies for cook...      0.00  
             Access to clean fuels and technologies for cook...      0.00  
             Access to electricity (% of population)                 0.00  
             Access to electricity, rural (% of rural popula...      0.00  
...                                                                   ...  
ZAF          Women who believe a husband is justified in bea...      0.00  
             Women who were first married by age 15 (% of wo...      0.15  
             Women who were first married by age 18 (% of wo...      2.15  
             Women's share of population ages 15+ living wit...      4.90  
             Young people (ages 15-24) newly infected with HIV   85000.00  

[59239 rows x 2 columns]

Once we got the quartiles, we compute the upper and lower limit, with a basic mathematical expression.

In [13]:
lower_limit=Q1 - 1.5 * IQR
lower=lower_limit.drop(['Date'],axis=1)
lower.rename(columns={"Value":"Lower limit"})

Lower limit
Country Code Indicator Name                                                 
ARE          Access to clean fuels and technologies for cook...      100.000
             Access to clean fuels and technologies for cook...      100.000
             Access to clean fuels and technologies for cook...      100.000
             Access to electricity (% of population)                 100.000
             Access to electricity, rural (% of rural popula...      100.000
...                                                                      ...
ZAF          Women who believe a husband is justified in bea...        1.000
             Women who were first married by age 15 (% of wo...        0.625
             Women who were first married by age 18 (% of wo...        1.375
             Women's share of population ages 15+ living wit...       49.850
             Young people (ages 15-24) newly infected with HIV    -22500.000

[59239 rows x 1 columns]

In [14]:
upper_limit=Q3 + 1.5 * IQR
upper=upper_limit.drop(['Date'],axis=1)
upper.rename(columns={"Value":"Upper limit"})

Upper limit
Country Code Indicator Name                                                 
ARE          Access to clean fuels and technologies for cook...      100.000
             Access to clean fuels and technologies for cook...      100.000
             Access to clean fuels and technologies for cook...      100.000
             Access to electricity (% of population)                 100.000
             Access to electricity, rural (% of rural popula...      100.000
...                                                                      ...
ZAF          Women who believe a husband is justified in bea...        1.000
             Women who were first married by age 15 (% of wo...        1.225
             Women who were first married by age 18 (% of wo...        9.975
             Women's share of population ages 15+ living wit...       69.450
             Young people (ages 15-24) newly infected with HIV    317500.000

[59239 rows x 1 columns]

Thirdly, we join the three tables we have (main dataframe, upper limit and lower limit) by matching country code and indicator name..

In [15]:
dfs = [BronzeDataFrame,lower,upper]
df_joined = ft.reduce(lambda left, right: pd.merge(left, right, on=['Country Code','Indicator Name']), dfs)
df_joined

Country Code                                     Indicator Name  Date  \
0                DZA  Access to clean fuels and technologies for coo...  2000   
1                DZA  Access to clean fuels and technologies for coo...  2001   
2                DZA  Access to clean fuels and technologies for coo...  2002   
3                DZA  Access to clean fuels and technologies for coo...  2003   
4                DZA  Access to clean fuels and technologies for coo...  2004   
...              ...                                                ...   ...   
1225413          YEM  Young people (ages 15-24) newly infected with HIV  2016   
1225414          YEM  Young people (ages 15-24) newly infected with HIV  2017   
1225415          YEM  Young people (ages 15-24) newly infected with HIV  2018   
1225416          YEM  Young people (ages 15-24) newly infected with HIV  2019   
1225417          YEM  Young people (ages 15-24) newly infected with HIV  2020   

         Value_x  Value_y  Value  
0           97.1     97.0  101.0  
1           97.3     97.0  101.0  
2           97.8     97.0  101.0  
3           98.0     97.0  101.0  
4           98.2     97.0  101.0  
...          ...      ...    ...  
1225413    200.0    -50.0  350.0  
1225414    200.0    -50.0  350.0  
1225415    200.0    -50.0  350.0  
1225416    200.0    -50.0  350.0  
1225417    200.0    -50.0  350.0  

[1225418 rows x 6 columns]

In [16]:
list(df_joined)

['Country Code', 'Indicator Name', 'Date', 'Value_x', 'Value_y', 'Value']

We rename the columns of the new table, as the columns headers are not saved after the joining. 

In [17]:
renamed=df_joined.set_axis(['Country','Indicator','Year', 'Real value', 'Lower value', 'Upper value'], axis=1, inplace=False)
renamed

Country                                          Indicator  Year  \
0           DZA  Access to clean fuels and technologies for coo...  2000   
1           DZA  Access to clean fuels and technologies for coo...  2001   
2           DZA  Access to clean fuels and technologies for coo...  2002   
3           DZA  Access to clean fuels and technologies for coo...  2003   
4           DZA  Access to clean fuels and technologies for coo...  2004   
...         ...                                                ...   ...   
1225413     YEM  Young people (ages 15-24) newly infected with HIV  2016   
1225414     YEM  Young people (ages 15-24) newly infected with HIV  2017   
1225415     YEM  Young people (ages 15-24) newly infected with HIV  2018   
1225416     YEM  Young people (ages 15-24) newly infected with HIV  2019   
1225417     YEM  Young people (ages 15-24) newly infected with HIV  2020   

         Real value  Lower value  Upper value  
0              97.1         97.0        101.0  
1              97.3         97.0        101.0  
2              97.8         97.0        101.0  
3              98.0         97.0        101.0  
4              98.2         97.0        101.0  
...             ...          ...          ...  
1225413       200.0        -50.0        350.0  
1225414       200.0        -50.0        350.0  
1225415       200.0        -50.0        350.0  
1225416       200.0        -50.0        350.0  
1225417       200.0        -50.0        350.0  

[1225418 rows x 6 columns]

Now that we have the table correctly defined, we remove from our dataframe the values that are outside our range, as it means that they are outliers.

In [18]:
sin_outliers=renamed.loc[~((renamed['Real value']<renamed['Lower value']) | (renamed['Real value']>renamed['Upper value']))]
sin_outliers

Country                                          Indicator  Year  \
0           DZA  Access to clean fuels and technologies for coo...  2000   
1           DZA  Access to clean fuels and technologies for coo...  2001   
2           DZA  Access to clean fuels and technologies for coo...  2002   
3           DZA  Access to clean fuels and technologies for coo...  2003   
4           DZA  Access to clean fuels and technologies for coo...  2004   
...         ...                                                ...   ...   
1225413     YEM  Young people (ages 15-24) newly infected with HIV  2016   
1225414     YEM  Young people (ages 15-24) newly infected with HIV  2017   
1225415     YEM  Young people (ages 15-24) newly infected with HIV  2018   
1225416     YEM  Young people (ages 15-24) newly infected with HIV  2019   
1225417     YEM  Young people (ages 15-24) newly infected with HIV  2020   

         Real value  Lower value  Upper value  
0              97.1         97.0        101.0  
1              97.3         97.0        101.0  
2              97.8         97.0        101.0  
3              98.0         97.0        101.0  
4              98.2         97.0        101.0  
...             ...          ...          ...  
1225413       200.0        -50.0        350.0  
1225414       200.0        -50.0        350.0  
1225415       200.0        -50.0        350.0  
1225416       200.0        -50.0        350.0  
1225417       200.0        -50.0        350.0  

[1189068 rows x 6 columns]

From the data above, we can perceive that our data comes down from 1225418  rows to 1189068, so 36.350  were outliers. The next steps are to order and display data better, removing those columns that we just do not need and pivoting the rows and columns. 

In [19]:
df_limpio=sin_outliers.drop(['Lower value','Upper value'],axis=1)
df_limpio

Country                                          Indicator  Year  \
0           DZA  Access to clean fuels and technologies for coo...  2000   
1           DZA  Access to clean fuels and technologies for coo...  2001   
2           DZA  Access to clean fuels and technologies for coo...  2002   
3           DZA  Access to clean fuels and technologies for coo...  2003   
4           DZA  Access to clean fuels and technologies for coo...  2004   
...         ...                                                ...   ...   
1225413     YEM  Young people (ages 15-24) newly infected with HIV  2016   
1225414     YEM  Young people (ages 15-24) newly infected with HIV  2017   
1225415     YEM  Young people (ages 15-24) newly infected with HIV  2018   
1225416     YEM  Young people (ages 15-24) newly infected with HIV  2019   
1225417     YEM  Young people (ages 15-24) newly infected with HIV  2020   

         Real value  
0              97.1  
1              97.3  
2              97.8  
3              98.0  
4              98.2  
...             ...  
1225413       200.0  
1225414       200.0  
1225415       200.0  
1225416       200.0  
1225417       200.0  

[1189068 rows x 4 columns]

In [20]:
cols=df_limpio['Indicator'].unique().tolist()

In [21]:
SilverDataFrame=df_limpio.set_index(["Country", "Year"]).pivot(columns="Indicator", values="Real value").reset_index()
SilverDataFrame

Indicator Country  Year  \
0             ARE  1990   
1             ARE  1991   
2             ARE  1992   
3             ARE  1993   
4             ARE  1994   
...           ...   ...   
1531          ZAF  2017   
1532          ZAF  2018   
1533          ZAF  2019   
1534          ZAF  2020   
1535          ZAF  2021   

Indicator  ARI treatment (% of children under 5 taken to a health provider)  \
0                                                        NaN                  
1                                                        NaN                  
2                                                        NaN                  
3                                                        NaN                  
4                                                        NaN                  
...                                                      ...                  
1531                                                     NaN                  
1532                                                     NaN                  
1533                                                     NaN                  
1534                                                     NaN                  
1535                                                     NaN                  

Indicator  Access to clean fuels and technologies for cooking (% of population)  \
0                                                        NaN                      
1                                                        NaN                      
2                                                        NaN                      
3                                                        NaN                      
4                                                        NaN                      
...                                                      ...                      
1531                                                    85.2                      
1532                                                    85.7                      
1533                                                    86.3                      
1534                                                    86.8                      
1535                                                     NaN                      

Indicator  Access to clean fuels and technologies for cooking, rural (% of rural population)  \
0                                                        NaN                                   
1                                                        NaN                                   
2                                                        NaN                                   
3                                                        NaN                                   
4                                                        NaN                                   
...                                                      ...                                   
1531                                                    64.6                                   
1532                                                    65.5                                   
1533                                                    65.5                                   
1534                                                    65.9                                   
1535                                                     NaN                                   

Indicator  Access to clean fuels and technologies for cooking, urban (% of urban population)  \
0                                                        NaN                                   
1                                                        NaN                                   
2                                                        NaN                                   
3                                                        NaN                                   
4                                                        NaN                                   


On the other hand, another big stone of normalizations is to nan/null values, which we have in all variables.

In [22]:
SilverDataFrame.isna().sum().sum()

1016628

As we can observe, we have lots of missing data, and as there is no optimal way to fullfill these values, thus, we will test some to arrive to the optimal method for our data set.

First, we need to create some lists so our loops work.

In [23]:
df=SilverDataFrame
europe_list=['DEU','FRA','SWE','GBR','ESP','HRV','POL','GRC','AUT','NLD']
persian_list=['IRQ','QAT','ARE','SAU','AZE','YEM','OMN']
naf_list=['DZA','EGY','LBY','ISR','TUR','MAR']
saf_list=['SEN','ZAF','LBR','MOZ','CMR','NGA','GHA']
asia_list=['BGD','IND','VNM','THA','IDN','PHL','KOR']
latam_list=['MEX','BRA','ARG','PER','VEN','COL','CHL','PAN','CRI']
two_list=['USA','CHN']
country_list=europe_list+persian_list+naf_list+saf_list+asia_list+latam_list+two_list


We are attempting the linear interpolation, which is achieved by geometrically rendering a straight line between two adjacent points on a graph or plane.

In [24]:
dat=df.loc[df.loc[:, 'Country'] == country_list[0]]
datc=dat.interpolate(method="linear")
data=datc

for i in range(1,len(country_list)):
    dat=df.loc[df.loc[:, 'Country'] == country_list[i]]
    datc=dat.interpolate(method="linear")
    data=pd.concat((data, datc), axis = 0)
data.isna().sum().sum()

685787

Here we attempt the backward filling, filling the previous cell with future values.

In [25]:
dat=df.loc[df.loc[:, 'Country'] == country_list[0]]
datc=dat.fillna(method='bfill')
data=datc

for i in range(1,len(country_list)):
    dat=df.loc[df.loc[:, 'Country'] == country_list[i]]
    datc=dat.fillna(method='bfill')
    data=pd.concat((data, datc), axis = 0)
data.isna().sum().sum()

498648

Here we will attempt the forward filling, which concists of filling the next cell with previous values.

In [26]:
dat=df.loc[df.loc[:, 'Country'] == country_list[0]]
datc=dat.fillna(method='ffill')
data=datc

for i in range(1,len(country_list)):
    dat=df.loc[df.loc[:, 'Country'] == country_list[i]]
    datc=dat.fillna(method='ffill')
    data=pd.concat((data, datc), axis = 0)
data.isna().sum().sum()

685787

The linear interpolation a form of interpolation, which involves the generation of new values based on an existing set of values. Linear interpolation is achieved by geometrically rendering a straight line between two adjacent points on a graph or plane. Whereas the backwards filling, will help us to arrive to those values which have not been fullfilled with the linear interpolation.

And as none of the methods have worked out correctly, independently, we are going to mix them, to achieve a better result.

In [27]:
dat=df.loc[df.loc[:, 'Country'] == country_list[0]]
datc=dat.interpolate(method="linear")
datc=datc.fillna(method='ffill')
data=datc

for i in range(1,len(country_list)):
    dat=df.loc[df.loc[:, 'Country'] == country_list[i]]
    datc=dat.interpolate(method="linear")
    datc=datc.fillna(method='ffill')
    data=pd.concat((data, datc), axis = 0)
data.isna().sum().sum()

685787

In [28]:
dat=df.loc[df.loc[:, 'Country'] == country_list[0]]
datc=dat.interpolate(method="linear")
datc=datc.fillna(method='bfill')
data=datc

for i in range(1,len(country_list)):
    dat=df.loc[df.loc[:, 'Country'] == country_list[i]]
    datc=dat.interpolate(method="linear")
    datc=datc.fillna(method='bfill')
    data=pd.concat((data, datc), axis = 0)
data.isna().sum().sum()

310048

And finally, mixing the three methods all together.

In [29]:
dat=df.loc[df.loc[:, 'Country'] == country_list[0]]
datc=dat.interpolate(method="linear")
datf=datc.fillna(method='bfill')
datr=datf.fillna(method='ffill')
data=datr

for i in range(1,len(country_list)):
    dat=df.loc[df.loc[:, 'Country'] == country_list[i]]
    datc=dat.interpolate(method="linear")
    datc=datc.fillna(method='bfill')
    datc=datc.fillna(method='ffill')
    data=pd.concat((data, datc), axis = 0)
data.isna().sum().sum()

310048

##### Conclusion

Therefore, the preferred method for the Nan values´ treatment that we are going to develop is a mix, between the linear interpolation and backwards filling.

In [30]:
dat=df.loc[df.loc[:, 'Country'] == country_list[0]]
datc=dat.interpolate(method="linear")
datf=datc.fillna(method='bfill')
datr=datf.fillna(method='ffill')
data=datr

for i in range(1,len(country_list)):
    dat=df.loc[df.loc[:, 'Country'] == country_list[i]]
    datc=dat.interpolate(method="linear")
    datc=datc.fillna(method='bfill')
    datc=datc.fillna(method='ffill')
    data=pd.concat((data, datc), axis = 0)
data

Indicator Country  Year  \
352           DEU  1990   
353           DEU  1991   
354           DEU  1992   
355           DEU  1993   
356           DEU  1994   
..            ...   ...   
251           CHN  2017   
252           CHN  2018   
253           CHN  2019   
254           CHN  2020   
255           CHN  2021   

Indicator  ARI treatment (% of children under 5 taken to a health provider)  \
352                                                      NaN                  
353                                                      NaN                  
354                                                      NaN                  
355                                                      NaN                  
356                                                      NaN                  
..                                                       ...                  
251                                                      NaN                  
252                                                      NaN                  
253                                                      NaN                  
254                                                      NaN                  
255                                                      NaN                  

Indicator  Access to clean fuels and technologies for cooking (% of population)  \
352                                                    100.0                      
353                                                    100.0                      
354                                                    100.0                      
355                                                    100.0                      
356                                                    100.0                      
..                                                       ...                      
251                                                     73.2                      
252                                                     75.6                      
253                                                     77.6                      
254                                                     79.4                      
255                                                     79.4                      

Indicator  Access to clean fuels and technologies for cooking, rural (% of rural population)  \
352                                                    100.0                                   
353                                                    100.0                                   
354                                                    100.0                                   
355                                                    100.0                                   
356                                                    100.0                                   
..                                                       ...                                   
251                                                     55.2                                   
252                                                     59.0                                   
253                                                     61.9                                   
254                                                     65.2                                   
255                                                     65.2                                   

Indicator  Access to clean fuels and technologies for cooking, urban (% of urban population)  \
352                                                    100.0                                   
353                                                    100.0                                   
354                                                    100.0                                   
355                                                    100.0                                   
356                                                    100.0                                   


Now, we will drop the columns which have over X% missing values because the absence of data creates an unreliable source. This % can be adjusted in the following slider. We have predetermined that 20% is a great starting point.

In [31]:
Slider1=widgets.FloatSlider(
    value=0.2,
    min=0,
    max=1.0,
    step=0.05,
    description='% that creates unreliable source:',
    disabled=False,
    continuous_update=False,
    orientation='horizontal',
    readout=True,
    readout_format='.1f',
)
Slider1

FloatSlider(value=0.2, continuous_update=False, description='% that creates unreliable source:', max=1.0, read…

In [32]:
number_1=len(data.index)*Slider1.value
for i in range(0, len(cols)):
    if data[cols[i]].isna().sum()>number_1:
        del(data[cols[i]])
        print(cols[i])
data

Adults (ages 15+) and children (ages 0-14) newly infected with HIV
Adults (ages 15-49) newly infected with HIV
Antiretroviral therapy coverage (% of people living with HIV)
Antiretroviral therapy coverage for PMTCT (% of pregnant women living with HIV)
ARI treatment (% of children under 5 taken to a health provider)
Average transaction cost of sending remittances to a specific country (%)
Average working hours of children, study and work, ages 7-14 (hours per week)
Average working hours of children, study and work, female, ages 7-14 (hours per week)
Average working hours of children, study and work, male, ages 7-14 (hours per week)
Average working hours of children, working only, ages 7-14 (hours per week)
Average working hours of children, working only, female, ages 7-14 (hours per week)
Average working hours of children, working only, male, ages 7-14 (hours per week)
Bank capital to assets ratio (%)
Bank liquid reserves to bank assets ratio (%)
Bank nonperforming loans to total gross

Indicator Country  Year  \
352           DEU  1990   
353           DEU  1991   
354           DEU  1992   
355           DEU  1993   
356           DEU  1994   
..            ...   ...   
251           CHN  2017   
252           CHN  2018   
253           CHN  2019   
254           CHN  2020   
255           CHN  2021   

Indicator  Access to clean fuels and technologies for cooking (% of population)  \
352                                                    100.0                      
353                                                    100.0                      
354                                                    100.0                      
355                                                    100.0                      
356                                                    100.0                      
..                                                       ...                      
251                                                     73.2                      
252                                                     75.6                      
253                                                     77.6                      
254                                                     79.4                      
255                                                     79.4                      

Indicator  Access to clean fuels and technologies for cooking, rural (% of rural population)  \
352                                                    100.0                                   
353                                                    100.0                                   
354                                                    100.0                                   
355                                                    100.0                                   
356                                                    100.0                                   
..                                                       ...                                   
251                                                     55.2                                   
252                                                     59.0                                   
253                                                     61.9                                   
254                                                     65.2                                   
255                                                     65.2                                   

Indicator  Access to clean fuels and technologies for cooking, urban (% of urban population)  \
352                                                    100.0                                   
353                                                    100.0                                   
354                                                    100.0                                   
355                                                    100.0                                   
356                                                    100.0                                   
..                                                       ...                                   
251                                                     86.2                                   
252                                                     87.4                                   
253                                                     88.4                                   
254                                                     89.4                                   
255                                                     89.4                                   

Indicator  Access to electricity (% of population)  \
352                                          100.0   
353                                          100.0   
354                                          100.0   
355                                          100.0   
356                                          100.0   
..                                             .

Afterwards, we have scaled the values. The escalation process has been done dividing each value by the initial one of an indicator (value in 1990). Considering the start point as 1 (initial value divided by itself), each result will show the growth respect to the initial data.

In [33]:
columns=data.columns.values.tolist()

In [34]:
datae=data.loc[data.loc[:, 'Country'] == country_list[0]]
for i in range(2,len(columns)):
    a=columns[i]
    datae[a]=datae[a]/datae.iloc[0,i]
datau=datae

In [35]:
for u in range(1,len(country_list)):
    datae=data.loc[data.loc[:, 'Country'] == country_list[u]]   
    for i in range(2,len(columns)):
        a=columns[i]
        datae[a]=datae[a]/datae.iloc[0,i]
    datau=pd.concat((datau, datae), axis = 0)
datau

Indicator Country  Year  \
352           DEU  1990   
353           DEU  1991   
354           DEU  1992   
355           DEU  1993   
356           DEU  1994   
..            ...   ...   
251           CHN  2017   
252           CHN  2018   
253           CHN  2019   
254           CHN  2020   
255           CHN  2021   

Indicator  Access to clean fuels and technologies for cooking (% of population)  \
352                                                 1.000000                      
353                                                 1.000000                      
354                                                 1.000000                      
355                                                 1.000000                      
356                                                 1.000000                      
..                                                       ...                      
251                                                 1.742857                      
252                                                 1.800000                      
253                                                 1.847619                      
254                                                 1.890476                      
255                                                 1.890476                      

Indicator  Access to clean fuels and technologies for cooking, rural (% of rural population)  \
352                                                 1.000000                                   
353                                                 1.000000                                   
354                                                 1.000000                                   
355                                                 1.000000                                   
356                                                 1.000000                                   
..                                                       ...                                   
251                                                 2.348936                                   
252                                                 2.510638                                   
253                                                 2.634043                                   
254                                                 2.774468                                   
255                                                 2.774468                                   

Indicator  Access to clean fuels and technologies for cooking, urban (% of urban population)  \
352                                                 1.000000                                   
353                                                 1.000000                                   
354                                                 1.000000                                   
355                                                 1.000000                                   
356                                                 1.000000                                   
..                                                       ...                                   
251                                                 1.261156                                   
252                                                 1.278713                                   
253                                                 1.293343                                   
254                                                 1.307974                                   
255                                                 1.307974                                   

Indicator  Access to electricity (% of population)  \
352                                       1.000000   
353                                       1.000000   
354                                       1.000000   
355                                       1.000000   
356                                       1.000000   
..                                             .

As later on we want to study the correlations with time moved, we need to create new columns for it. The reason why is because, maybe the effect of a variable does not happen until a couple of years later. The time movements that have been considered are those of the Fibonacci serie within our time period. 

The following pictures helps to realize the behaviour that we were explaining. 

![](https://raw.githubusercontent.com/devonfw-forge/python-data-driven-decisions/main-the-big-three/Logos/Time%20moved.JPG)

In [36]:
shifted=pd.DataFrame()
for i in range(0,len(country_list)):
    dat=datau.loc[datau.loc[:, 'Country'] == country_list[i]]
    dat['GDP (current US$)+1']=dat['GDP (current US$)'].shift(periods=1)
    dat['GDP (current US$)+2']=dat['GDP (current US$)'].shift(periods=2)
    dat['GDP (current US$)+3']=dat['GDP (current US$)'].shift(periods=3)
    dat['GDP (current US$)+5']=dat['GDP (current US$)'].shift(periods=5)
    dat['GDP (current US$)+8']=dat['GDP (current US$)'].shift(periods=8)
    dat['GDP (current US$)+13']=dat['GDP (current US$)'].shift(periods=13)
    dat['GDP (current US$)+21']=dat['GDP (current US$)'].shift(periods=21)
    shifted=pd.concat((shifted, dat), axis = 0)
shifted

Indicator Country  Year  \
352           DEU  1990   
353           DEU  1991   
354           DEU  1992   
355           DEU  1993   
356           DEU  1994   
..            ...   ...   
251           CHN  2017   
252           CHN  2018   
253           CHN  2019   
254           CHN  2020   
255           CHN  2021   

Indicator  Access to clean fuels and technologies for cooking (% of population)  \
352                                                 1.000000                      
353                                                 1.000000                      
354                                                 1.000000                      
355                                                 1.000000                      
356                                                 1.000000                      
..                                                       ...                      
251                                                 1.742857                      
252                                                 1.800000                      
253                                                 1.847619                      
254                                                 1.890476                      
255                                                 1.890476                      

Indicator  Access to clean fuels and technologies for cooking, rural (% of rural population)  \
352                                                 1.000000                                   
353                                                 1.000000                                   
354                                                 1.000000                                   
355                                                 1.000000                                   
356                                                 1.000000                                   
..                                                       ...                                   
251                                                 2.348936                                   
252                                                 2.510638                                   
253                                                 2.634043                                   
254                                                 2.774468                                   
255                                                 2.774468                                   

Indicator  Access to clean fuels and technologies for cooking, urban (% of urban population)  \
352                                                 1.000000                                   
353                                                 1.000000                                   
354                                                 1.000000                                   
355                                                 1.000000                                   
356                                                 1.000000                                   
..                                                       ...                                   
251                                                 1.261156                                   
252                                                 1.278713                                   
253                                                 1.293343                                   
254                                                 1.307974                                   
255                                                 1.307974                                   

Indicator  Access to electricity (% of population)  \
352                                       1.000000   
353                                       1.000000   
354                                       1.000000   
355                                       1.000000   
356                                       1.000000   
..                                             .

In [37]:
data=shifted

For the next part of analyzing this data, we think it is gonna be interesting to have it classify by the categories of the Country groups defined before, to which we call "Continent". This category is useful as it groups the nations with similar economies or geographical proximity, so we can extract common conclusions from them.

We create a dictionary with the regions and the countries included in each one. Where we will relate the countries and regions so then we can apply the .map function and arrive to the final dataframe.

In [38]:
countries_by_region = {
    "Europe": ('DEU','FRA','SWE','GBR','ESP','HRV','POL','GRC','AUT','NLD'),
    'Persian Gulf': ('IRQ','QAT','ARE','SAU','AZE','YEM','YDR','OMN'),
    'North Africa':('DZA','EGY','LBY','ISR','TUR','MAR'),
    'South Africa':('SEN','ZAF','LBR','MOZ','CMR','NGA','GHA'),
    'Asia':('BGD','IND','VNM','THA','IDN','PHL','KOR'),
    'Latam':('MEX','BRA','ARG','PER','VEN','COL','CHL','PAN','CRI'),
    'Pair':('USA','CHN')
    }

all_countries = {}
for region in countries_by_region.keys():
  for country in countries_by_region[region]:
    all_countries[country] = region

print(all_countries)

{'DEU': 'Europe', 'FRA': 'Europe', 'SWE': 'Europe', 'GBR': 'Europe', 'ESP': 'Europe', 'HRV': 'Europe', 'POL': 'Europe', 'GRC': 'Europe', 'AUT': 'Europe', 'NLD': 'Europe', 'IRQ': 'Persian Gulf', 'QAT': 'Persian Gulf', 'ARE': 'Persian Gulf', 'SAU': 'Persian Gulf', 'AZE': 'Persian Gulf', 'YEM': 'Persian Gulf', 'YDR': 'Persian Gulf', 'OMN': 'Persian Gulf', 'DZA': 'North Africa', 'EGY': 'North Africa', 'LBY': 'North Africa', 'ISR': 'North Africa', 'TUR': 'North Africa', 'MAR': 'North Africa', 'SEN': 'South Africa', 'ZAF': 'South Africa', 'LBR': 'South Africa', 'MOZ': 'South Africa', 'CMR': 'South Africa', 'NGA': 'South Africa', 'GHA': 'South Africa', 'BGD': 'Asia', 'IND': 'Asia', 'VNM': 'Asia', 'THA': 'Asia', 'IDN': 'Asia', 'PHL': 'Asia', 'KOR': 'Asia', 'MEX': 'Latam', 'BRA': 'Latam', 'ARG': 'Latam', 'PER': 'Latam', 'VEN': 'Latam', 'COL': 'Latam', 'CHL': 'Latam', 'PAN': 'Latam', 'CRI': 'Latam', 'USA': 'Pair', 'CHN': 'Pair'}


In [39]:
data['Continent']=data['Country'].map(all_countries)
Goldendataframe=data
Goldendataframe

Indicator Country  Year  \
352           DEU  1990   
353           DEU  1991   
354           DEU  1992   
355           DEU  1993   
356           DEU  1994   
..            ...   ...   
251           CHN  2017   
252           CHN  2018   
253           CHN  2019   
254           CHN  2020   
255           CHN  2021   

Indicator  Access to clean fuels and technologies for cooking (% of population)  \
352                                                 1.000000                      
353                                                 1.000000                      
354                                                 1.000000                      
355                                                 1.000000                      
356                                                 1.000000                      
..                                                       ...                      
251                                                 1.742857                      
252                                                 1.800000                      
253                                                 1.847619                      
254                                                 1.890476                      
255                                                 1.890476                      

Indicator  Access to clean fuels and technologies for cooking, rural (% of rural population)  \
352                                                 1.000000                                   
353                                                 1.000000                                   
354                                                 1.000000                                   
355                                                 1.000000                                   
356                                                 1.000000                                   
..                                                       ...                                   
251                                                 2.348936                                   
252                                                 2.510638                                   
253                                                 2.634043                                   
254                                                 2.774468                                   
255                                                 2.774468                                   

Indicator  Access to clean fuels and technologies for cooking, urban (% of urban population)  \
352                                                 1.000000                                   
353                                                 1.000000                                   
354                                                 1.000000                                   
355                                                 1.000000                                   
356                                                 1.000000                                   
..                                                       ...                                   
251                                                 1.261156                                   
252                                                 1.278713                                   
253                                                 1.293343                                   
254                                                 1.307974                                   
255                                                 1.307974                                   

Indicator  Access to electricity (% of population)  \
352                                       1.000000   
353                                       1.000000   
354                                       1.000000   
355                                       1.000000   
356                                       1.000000   
..                                             .

With that all, we export our dataframe all-in-one and by the continent category.

In [40]:
Goldendataframe.to_csv(os.getcwd()+'/Data/GoldenDataFrame.csv')

In [41]:
for region, data in Goldendataframe.groupby('Continent'):
   data.to_csv(os.getcwd()+'/Data/{}.csv'.format(region))

# Categorization of variables

In this section, we are going to attempt a categorization of the whole of the variables, which most of them come the same sources and just differ in the units that are measured, or the total that they are refering, between others. For a simpler treatment of the data, the variables have been pivoted into the same column.

In [42]:
columns_golden=list(Goldendataframe.columns)
del columns_golden[0:2]

In [43]:
Categorization=Goldendataframe.set_index(['Country','Year', 'Continent']).stack().reset_index()
Categorization['Short indicator']=Categorization['Indicator']
Categorization

Country  Year Continent  \
0           DEU  1990    Europe   
1           DEU  1990    Europe   
2           DEU  1990    Europe   
3           DEU  1990    Europe   
4           DEU  1990    Europe   
...         ...   ...       ...   
1540358     CHN  2021      Pair   
1540359     CHN  2021      Pair   
1540360     CHN  2021      Pair   
1540361     CHN  2021      Pair   
1540362     CHN  2021      Pair   

                                                 Indicator          0  \
0        Access to clean fuels and technologies for coo...   1.000000   
1        Access to clean fuels and technologies for coo...   1.000000   
2        Access to clean fuels and technologies for coo...   1.000000   
3                  Access to electricity (% of population)   1.000000   
4        Access to electricity, rural (% of rural popul...   1.000000   
...                                                    ...        ...   
1540358                                GDP (current US$)+3  38.504955   
1540359                                GDP (current US$)+5  31.129362   
1540360                                GDP (current US$)+8  26.521259   
1540361                               GDP (current US$)+13  12.731623   
1540362                               GDP (current US$)+21   3.356853   

                                           Short indicator  
0        Access to clean fuels and technologies for coo...  
1        Access to clean fuels and technologies for coo...  
2        Access to clean fuels and technologies for coo...  
3                  Access to electricity (% of population)  
4        Access to electricity, rural (% of rural popul...  
...                                                    ...  
1540358                                GDP (current US$)+3  
1540359                                GDP (current US$)+5  
1540360                                GDP (current US$)+8  
1540361                               GDP (current US$)+13  
1540362                               GDP (current US$)+21  

[1540363 rows x 6 columns]

 There are some indicators which represent exactly the same through different units, so, we are going to select only one type. For example, in monetary cases, indicators which are expressed with current US $ has been selected. Then, which are showed with the percentage and the total value, we have programmed to selct which ones which show a greater value.

The links used to learn about these functions have been:

https://www.geeksforgeeks.org/how-to-drop-rows-that-contain-a-specific-string-in-pandas/ 

https://www.statology.org/pandas-drop-rows-that-contain-string/ 

In [44]:
import re
discard=["annual % growth","constant 2015 US[$]","% of GNI","constant LCU","current LCU"]
Categorization2=Categorization[~Categorization['Short indicator'].str.contains('|'.join(discard))]
Categorization2

Country  Year Continent  \
0           DEU  1990    Europe   
1           DEU  1990    Europe   
2           DEU  1990    Europe   
3           DEU  1990    Europe   
4           DEU  1990    Europe   
...         ...   ...       ...   
1540358     CHN  2021      Pair   
1540359     CHN  2021      Pair   
1540360     CHN  2021      Pair   
1540361     CHN  2021      Pair   
1540362     CHN  2021      Pair   

                                                 Indicator          0  \
0        Access to clean fuels and technologies for coo...   1.000000   
1        Access to clean fuels and technologies for coo...   1.000000   
2        Access to clean fuels and technologies for coo...   1.000000   
3                  Access to electricity (% of population)   1.000000   
4        Access to electricity, rural (% of rural popul...   1.000000   
...                                                    ...        ...   
1540358                                GDP (current US$)+3  38.504955   
1540359                                GDP (current US$)+5  31.129362   
1540360                                GDP (current US$)+8  26.521259   
1540361                               GDP (current US$)+13  12.731623   
1540362                               GDP (current US$)+21   3.356853   

                                           Short indicator  
0        Access to clean fuels and technologies for coo...  
1        Access to clean fuels and technologies for coo...  
2        Access to clean fuels and technologies for coo...  
3                  Access to electricity (% of population)  
4        Access to electricity, rural (% of rural popul...  
...                                                    ...  
1540358                                GDP (current US$)+3  
1540359                                GDP (current US$)+5  
1540360                                GDP (current US$)+8  
1540361                               GDP (current US$)+13  
1540362                               GDP (current US$)+21  

[1372098 rows x 6 columns]

To check previous step.

In [45]:
#Categorization2.apply(lambda row: row.astype(str).str.contains('US').any(), axis=1)

Now we are going to structure the indicators in a same way to work better. The first step consist of making a new column that shows the units of each variable. Units are showed inside the parenthesis of the indicator name.

In [46]:
Categorization2['Units']=Categorization2['Short indicator'].str.extract(' (\(.*\))')
Categorization2

Country  Year Continent  \
0           DEU  1990    Europe   
1           DEU  1990    Europe   
2           DEU  1990    Europe   
3           DEU  1990    Europe   
4           DEU  1990    Europe   
...         ...   ...       ...   
1540358     CHN  2021      Pair   
1540359     CHN  2021      Pair   
1540360     CHN  2021      Pair   
1540361     CHN  2021      Pair   
1540362     CHN  2021      Pair   

                                                 Indicator          0  \
0        Access to clean fuels and technologies for coo...   1.000000   
1        Access to clean fuels and technologies for coo...   1.000000   
2        Access to clean fuels and technologies for coo...   1.000000   
3                  Access to electricity (% of population)   1.000000   
4        Access to electricity, rural (% of rural popul...   1.000000   
...                                                    ...        ...   
1540358                                GDP (current US$)+3  38.504955   
1540359                                GDP (current US$)+5  31.129362   
1540360                                GDP (current US$)+8  26.521259   
1540361                               GDP (current US$)+13  12.731623   
1540362                               GDP (current US$)+21   3.356853   

                                           Short indicator  \
0        Access to clean fuels and technologies for coo...   
1        Access to clean fuels and technologies for coo...   
2        Access to clean fuels and technologies for coo...   
3                  Access to electricity (% of population)   
4        Access to electricity, rural (% of rural popul...   
...                                                    ...   
1540358                                GDP (current US$)+3   
1540359                                GDP (current US$)+5   
1540360                                GDP (current US$)+8   
1540361                               GDP (current US$)+13   
1540362                               GDP (current US$)+21   

                           Units  
0              (% of population)  
1        (% of rural population)  
2        (% of urban population)  
3              (% of population)  
4        (% of rural population)  
...                          ...  
1540358            (current US$)  
1540359            (current US$)  
1540360            (current US$)  
1540361            (current US$)  
1540362            (current US$)  

[1372098 rows x 7 columns]

Now, short indicator refers to the original indicator name without the units. The extracted information from the origin column has been deleted.

In [47]:
Categorization2['Short indicator']=Categorization2['Short indicator'].str.replace(r" (\(.*\))","")
Categorization2

Country  Year Continent  \
0           DEU  1990    Europe   
1           DEU  1990    Europe   
2           DEU  1990    Europe   
3           DEU  1990    Europe   
4           DEU  1990    Europe   
...         ...   ...       ...   
1540358     CHN  2021      Pair   
1540359     CHN  2021      Pair   
1540360     CHN  2021      Pair   
1540361     CHN  2021      Pair   
1540362     CHN  2021      Pair   

                                                 Indicator          0  \
0        Access to clean fuels and technologies for coo...   1.000000   
1        Access to clean fuels and technologies for coo...   1.000000   
2        Access to clean fuels and technologies for coo...   1.000000   
3                  Access to electricity (% of population)   1.000000   
4        Access to electricity, rural (% of rural popul...   1.000000   
...                                                    ...        ...   
1540358                                GDP (current US$)+3  38.504955   
1540359                                GDP (current US$)+5  31.129362   
1540360                                GDP (current US$)+8  26.521259   
1540361                               GDP (current US$)+13  12.731623   
1540362                               GDP (current US$)+21   3.356853   

                                           Short indicator  \
0        Access to clean fuels and technologies for coo...   
1        Access to clean fuels and technologies for coo...   
2        Access to clean fuels and technologies for coo...   
3                                    Access to electricity   
4                             Access to electricity, rural   
...                                                    ...   
1540358                                              GDP+3   
1540359                                              GDP+5   
1540360                                              GDP+8   
1540361                                             GDP+13   
1540362                                             GDP+21   

                           Units  
0              (% of population)  
1        (% of rural population)  
2        (% of urban population)  
3              (% of population)  
4        (% of rural population)  
...                          ...  
1540358            (current US$)  
1540359            (current US$)  
1540360            (current US$)  
1540361            (current US$)  
1540362            (current US$)  

[1372098 rows x 7 columns]

In some cases there are extra information in indicators name. The information of the second parenthesis is extracted as a new column too.

In [48]:
two_parent=Categorization2[Categorization2['Short indicator'].str.contains('Contributing family workers')]
two_parent

Country  Year Continent  \
151         DEU  1990    Europe   
152         DEU  1990    Europe   
153         DEU  1990    Europe   
1151        DEU  1991    Europe   
1152        DEU  1991    Europe   
...         ...   ...       ...   
1538535     CHN  2020      Pair   
1538536     CHN  2020      Pair   
1539524     CHN  2021      Pair   
1539525     CHN  2021      Pair   
1539526     CHN  2021      Pair   

                                                 Indicator         0  \
151      Contributing family workers, female (% of fema...  1.000000   
152      Contributing family workers, male (% of male e...  1.000000   
153      Contributing family workers, total (% of total...  1.000000   
1151     Contributing family workers, female (% of fema...  1.000000   
1152     Contributing family workers, male (% of male e...  1.000000   
...                                                    ...       ...   
1538535  Contributing family workers, male (% of male e...  0.252894   
1538536  Contributing family workers, total (% of total...  0.329385   
1539524  Contributing family workers, female (% of fema...  0.386565   
1539525  Contributing family workers, male (% of male e...  0.252894   
1539526  Contributing family workers, total (% of total...  0.329385   

                             Short indicator  \
151      Contributing family workers, female   
152        Contributing family workers, male   
153       Contributing family workers, total   
1151     Contributing family workers, female   
1152       Contributing family workers, male   
...                                      ...   
1538535    Contributing family workers, male   
1538536   Contributing family workers, total   
1539524  Contributing family workers, female   
1539525    Contributing family workers, male   
1539526   Contributing family workers, total   

                                                   Units  
151      (% of female employment) (modeled ILO estimate)  
152        (% of male employment) (modeled ILO estimate)  
153       (% of total employment) (modeled ILO estimate)  
1151     (% of female employment) (modeled ILO estimate)  
1152       (% of male employment) (modeled ILO estimate)  
...                                                  ...  
1538535    (% of male employment) (modeled ILO estimate)  
1538536   (% of total employment) (modeled ILO estimate)  
1539524  (% of female employment) (modeled ILO estimate)  
1539525    (% of male employment) (modeled ILO estimate)  
1539526   (% of total employment) (modeled ILO estimate)  

[4545 rows x 7 columns]

Moreover, there are some inidcators with an extra parenthesis adding some  more information. As this information isn't related with units, another column named as 'other specification' has been created.

In [49]:
Categorization2[['Units','Other specification']]=Categorization2['Units'].str.split("\) ", n=1,expand=True)
Categorization2

Country  Year Continent  \
0           DEU  1990    Europe   
1           DEU  1990    Europe   
2           DEU  1990    Europe   
3           DEU  1990    Europe   
4           DEU  1990    Europe   
...         ...   ...       ...   
1540358     CHN  2021      Pair   
1540359     CHN  2021      Pair   
1540360     CHN  2021      Pair   
1540361     CHN  2021      Pair   
1540362     CHN  2021      Pair   

                                                 Indicator          0  \
0        Access to clean fuels and technologies for coo...   1.000000   
1        Access to clean fuels and technologies for coo...   1.000000   
2        Access to clean fuels and technologies for coo...   1.000000   
3                  Access to electricity (% of population)   1.000000   
4        Access to electricity, rural (% of rural popul...   1.000000   
...                                                    ...        ...   
1540358                                GDP (current US$)+3  38.504955   
1540359                                GDP (current US$)+5  31.129362   
1540360                                GDP (current US$)+8  26.521259   
1540361                               GDP (current US$)+13  12.731623   
1540362                               GDP (current US$)+21   3.356853   

                                           Short indicator  \
0        Access to clean fuels and technologies for coo...   
1        Access to clean fuels and technologies for coo...   
2        Access to clean fuels and technologies for coo...   
3                                    Access to electricity   
4                             Access to electricity, rural   
...                                                    ...   
1540358                                              GDP+3   
1540359                                              GDP+5   
1540360                                              GDP+8   
1540361                                             GDP+13   
1540362                                             GDP+21   

                           Units Other specification  
0              (% of population)                None  
1        (% of rural population)                None  
2        (% of urban population)                None  
3              (% of population)                None  
4        (% of rural population)                None  
...                          ...                 ...  
1540358            (current US$)                None  
1540359            (current US$)                None  
1540360            (current US$)                None  
1540361            (current US$)                None  
1540362            (current US$)                None  

[1372098 rows x 8 columns]

At the end of the variable name, separated by the last "," it is informing us about to which subgroup makes reference the variable. Thus, there are some indicators that have information divided for small groups. This information is shown as a new column named 'Subgroup'.

In [50]:
Categorization2[['Subgroup']]=Categorization2['Short indicator'].str.extract(',(?P<field>[^,]*?)$')
Categorization2

Country  Year Continent  \
0           DEU  1990    Europe   
1           DEU  1990    Europe   
2           DEU  1990    Europe   
3           DEU  1990    Europe   
4           DEU  1990    Europe   
...         ...   ...       ...   
1540358     CHN  2021      Pair   
1540359     CHN  2021      Pair   
1540360     CHN  2021      Pair   
1540361     CHN  2021      Pair   
1540362     CHN  2021      Pair   

                                                 Indicator          0  \
0        Access to clean fuels and technologies for coo...   1.000000   
1        Access to clean fuels and technologies for coo...   1.000000   
2        Access to clean fuels and technologies for coo...   1.000000   
3                  Access to electricity (% of population)   1.000000   
4        Access to electricity, rural (% of rural popul...   1.000000   
...                                                    ...        ...   
1540358                                GDP (current US$)+3  38.504955   
1540359                                GDP (current US$)+5  31.129362   
1540360                                GDP (current US$)+8  26.521259   
1540361                               GDP (current US$)+13  12.731623   
1540362                               GDP (current US$)+21   3.356853   

                                           Short indicator  \
0        Access to clean fuels and technologies for coo...   
1        Access to clean fuels and technologies for coo...   
2        Access to clean fuels and technologies for coo...   
3                                    Access to electricity   
4                             Access to electricity, rural   
...                                                    ...   
1540358                                              GDP+3   
1540359                                              GDP+5   
1540360                                              GDP+8   
1540361                                             GDP+13   
1540362                                             GDP+21   

                           Units Other specification Subgroup  
0              (% of population)                None      NaN  
1        (% of rural population)                None    rural  
2        (% of urban population)                None    urban  
3              (% of population)                None      NaN  
4        (% of rural population)                None    rural  
...                          ...                 ...      ...  
1540358            (current US$)                None      NaN  
1540359            (current US$)                None      NaN  
1540360            (current US$)                None      NaN  
1540361            (current US$)                None      NaN  
1540362            (current US$)                None      NaN  

[1372098 rows x 9 columns]

As before, information which is shown as a new column is deleted from the origin one.

In [51]:
Categorization2['Short indicator']=Categorization2['Short indicator'].str.replace(',(?P<field>[^,]*?)$',"")
Categorization2

Country  Year Continent  \
0           DEU  1990    Europe   
1           DEU  1990    Europe   
2           DEU  1990    Europe   
3           DEU  1990    Europe   
4           DEU  1990    Europe   
...         ...   ...       ...   
1540358     CHN  2021      Pair   
1540359     CHN  2021      Pair   
1540360     CHN  2021      Pair   
1540361     CHN  2021      Pair   
1540362     CHN  2021      Pair   

                                                 Indicator          0  \
0        Access to clean fuels and technologies for coo...   1.000000   
1        Access to clean fuels and technologies for coo...   1.000000   
2        Access to clean fuels and technologies for coo...   1.000000   
3                  Access to electricity (% of population)   1.000000   
4        Access to electricity, rural (% of rural popul...   1.000000   
...                                                    ...        ...   
1540358                                GDP (current US$)+3  38.504955   
1540359                                GDP (current US$)+5  31.129362   
1540360                                GDP (current US$)+8  26.521259   
1540361                               GDP (current US$)+13  12.731623   
1540362                               GDP (current US$)+21   3.356853   

                                           Short indicator  \
0        Access to clean fuels and technologies for coo...   
1        Access to clean fuels and technologies for coo...   
2        Access to clean fuels and technologies for coo...   
3                                    Access to electricity   
4                                    Access to electricity   
...                                                    ...   
1540358                                              GDP+3   
1540359                                              GDP+5   
1540360                                              GDP+8   
1540361                                             GDP+13   
1540362                                             GDP+21   

                           Units Other specification Subgroup  
0              (% of population)                None      NaN  
1        (% of rural population)                None    rural  
2        (% of urban population)                None    urban  
3              (% of population)                None      NaN  
4        (% of rural population)                None    rural  
...                          ...                 ...      ...  
1540358            (current US$)                None      NaN  
1540359            (current US$)                None      NaN  
1540360            (current US$)                None      NaN  
1540361            (current US$)                None      NaN  
1540362            (current US$)                None      NaN  

[1372098 rows x 9 columns]

All the indicators don't have these elements. So, a checking point is needed.

In [52]:
Categorization2['Subgroup']=Categorization2['Subgroup'].replace(['None'],['total'])
Categorization2['Subgroup']=Categorization2['Subgroup'].fillna('total')

There are some duplicate variables which should be removed too.

In [53]:
Categorization2.drop_duplicates(subset=['Country','Year','Short indicator','Continent','Subgroup'], keep='first')

Country  Year Continent  \
0           DEU  1990    Europe   
1           DEU  1990    Europe   
2           DEU  1990    Europe   
3           DEU  1990    Europe   
4           DEU  1990    Europe   
...         ...   ...       ...   
1540358     CHN  2021      Pair   
1540359     CHN  2021      Pair   
1540360     CHN  2021      Pair   
1540361     CHN  2021      Pair   
1540362     CHN  2021      Pair   

                                                 Indicator          0  \
0        Access to clean fuels and technologies for coo...   1.000000   
1        Access to clean fuels and technologies for coo...   1.000000   
2        Access to clean fuels and technologies for coo...   1.000000   
3                  Access to electricity (% of population)   1.000000   
4        Access to electricity, rural (% of rural popul...   1.000000   
...                                                    ...        ...   
1540358                                GDP (current US$)+3  38.504955   
1540359                                GDP (current US$)+5  31.129362   
1540360                                GDP (current US$)+8  26.521259   
1540361                               GDP (current US$)+13  12.731623   
1540362                               GDP (current US$)+21   3.356853   

                                           Short indicator  \
0        Access to clean fuels and technologies for coo...   
1        Access to clean fuels and technologies for coo...   
2        Access to clean fuels and technologies for coo...   
3                                    Access to electricity   
4                                    Access to electricity   
...                                                    ...   
1540358                                              GDP+3   
1540359                                              GDP+5   
1540360                                              GDP+8   
1540361                                             GDP+13   
1540362                                             GDP+21   

                           Units Other specification Subgroup  
0              (% of population)                None    total  
1        (% of rural population)                None    rural  
2        (% of urban population)                None    urban  
3              (% of population)                None    total  
4        (% of rural population)                None    rural  
...                          ...                 ...      ...  
1540358            (current US$)                None    total  
1540359            (current US$)                None    total  
1540360            (current US$)                None    total  
1540361            (current US$)                None    total  
1540362            (current US$)                None    total  

[1157276 rows x 9 columns]

Reordering columns, categorization3 is our df after all these division in categories.

In [54]:
Categorization2.rename(columns={Categorization2.columns[4]:'Value'},inplace=True)
Categorization3=Categorization2[['Country','Year','Continent','Indicator','Short indicator','Value','Subgroup','Units','Other specification']]
Categorization3

Country  Year Continent  \
0           DEU  1990    Europe   
1           DEU  1990    Europe   
2           DEU  1990    Europe   
3           DEU  1990    Europe   
4           DEU  1990    Europe   
...         ...   ...       ...   
1540358     CHN  2021      Pair   
1540359     CHN  2021      Pair   
1540360     CHN  2021      Pair   
1540361     CHN  2021      Pair   
1540362     CHN  2021      Pair   

                                                 Indicator  \
0        Access to clean fuels and technologies for coo...   
1        Access to clean fuels and technologies for coo...   
2        Access to clean fuels and technologies for coo...   
3                  Access to electricity (% of population)   
4        Access to electricity, rural (% of rural popul...   
...                                                    ...   
1540358                                GDP (current US$)+3   
1540359                                GDP (current US$)+5   
1540360                                GDP (current US$)+8   
1540361                               GDP (current US$)+13   
1540362                               GDP (current US$)+21   

                                           Short indicator      Value  \
0        Access to clean fuels and technologies for coo...   1.000000   
1        Access to clean fuels and technologies for coo...   1.000000   
2        Access to clean fuels and technologies for coo...   1.000000   
3                                    Access to electricity   1.000000   
4                                    Access to electricity   1.000000   
...                                                    ...        ...   
1540358                                              GDP+3  38.504955   
1540359                                              GDP+5  31.129362   
1540360                                              GDP+8  26.521259   
1540361                                             GDP+13  12.731623   
1540362                                             GDP+21   3.356853   

        Subgroup                    Units Other specification  
0          total        (% of population)                None  
1          rural  (% of rural population)                None  
2          urban  (% of urban population)                None  
3          total        (% of population)                None  
4          rural  (% of rural population)                None  
...          ...                      ...                 ...  
1540358    total            (current US$)                None  
1540359    total            (current US$)                None  
1540360    total            (current US$)                None  
1540361    total            (current US$)                None  
1540362    total            (current US$)                None  

[1372098 rows x 9 columns]

------------------------------

In [55]:
Categorization4=Categorization3.loc[Categorization3['Subgroup']=='total']
Categorization4

Country  Year Continent  \
0           DEU  1990    Europe   
3           DEU  1990    Europe   
6           DEU  1990    Europe   
20          DEU  1990    Europe   
23          DEU  1990    Europe   
...         ...   ...       ...   
1540358     CHN  2021      Pair   
1540359     CHN  2021      Pair   
1540360     CHN  2021      Pair   
1540361     CHN  2021      Pair   
1540362     CHN  2021      Pair   

                                                 Indicator  \
0        Access to clean fuels and technologies for coo...   
3                  Access to electricity (% of population)   
6        Account ownership at a financial institution o...   
20              Adjusted net national income (current US$)   
23       Adjusted net national income per capita (curre...   
...                                                    ...   
1540358                                GDP (current US$)+3   
1540359                                GDP (current US$)+5   
1540360                                GDP (current US$)+8   
1540361                               GDP (current US$)+13   
1540362                               GDP (current US$)+21   

                                           Short indicator      Value  \
0        Access to clean fuels and technologies for coo...   1.000000   
3                                    Access to electricity   1.000000   
6        Account ownership at a financial institution o...   1.000000   
20                            Adjusted net national income   1.000000   
23                 Adjusted net national income per capita   1.000000   
...                                                    ...        ...   
1540358                                              GDP+3  38.504955   
1540359                                              GDP+5  31.129362   
1540360                                              GDP+8  26.521259   
1540361                                             GDP+13  12.731623   
1540362                                             GDP+21   3.356853   

        Subgroup                       Units Other specification  
0          total           (% of population)                None  
3          total           (% of population)                None  
6          total  (% of population ages 15+)                None  
20         total               (current US$)                None  
23         total               (current US$)                None  
...          ...                         ...                 ...  
1540358    total               (current US$)                None  
1540359    total               (current US$)                None  
1540360    total               (current US$)                None  
1540361    total               (current US$)                None  
1540362    total               (current US$)                None  

[687830 rows x 9 columns]

In [56]:
Categorization4.to_csv(os.getcwd()+'/Data/Categorization.csv')